compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -671,22 +672,34 @@ public void visit(BLangWhile whileNode, AnalyzerData data) {
                 data.breakAsLastStatement || data.statementReturnsPanicsOrFails, true,
                 data.potentiallyInvalidAssignmentInLoopsInfo);
 
-        decrementLoopCount(data);
         data.failureHandled = failureHandled;
-        if (data.booleanConstCondition != symTable.trueType || data.breakStmtFound) {
+        if (isUnreachableBlock(whileNode, data)) {
+            data.infiniteLoop = true;
             data.statementReturnsPanicsOrFails = prevStatementReturnsPanicsOrFails;
             data.continueAsLastStatement = prevContinueAsLastStatement;
             data.breakAsLastStatement = prevBreakAsLastStatement;
+        } else if (data.booleanConstCondition != symTable.trueType || data.breakStmtFound) {
+            data.statementReturnsPanicsOrFails = prevStatementReturnsPanicsOrFails;
+            data.continueAsLastStatement = prevContinueAsLastStatement;
+            data.breakAsLastStatement = prevBreakAsLastStatement;
+            data.infiniteLoop = previousInfiniteLoop;
         } else {
             data.statementReturnsPanicsOrFails = true;
```

Comment:
IMO, this is not correct. The statement returns panic or fails check when visiting the body of the while loop. When visiting a `BLangFail` node or a `BLangPanic` node, we need to set this field. Here, when visiting the `BLangWhile` node, what we need to do is check whether the loop is infinite or not.